### PR TITLE
[PPWM-83][PPWM-84][PPWM-85][PPWM-86] Migration fixes after testing in production

### PR DESCRIPTION
**Pull request by ehigham**
[PPWM-83] Conditionally set requester pays
RR: https://broadworkbench.atlassian.net/browse/PPWM-83
Only set the bucket requester pays flag when we need to. This helps
avoid some failures when billing accounts are deativated or "delinquent"
(as google puts it).

[PPWM-84] Don't set billing account when reusing google project
RR: https://broadworkbench.atlassian.net/browse/PPWM-84
Turns out this is not a noop when setting the billing account to the
same value and we've run into quota issues in prod because of this.

[PPWM-85] Split transfer job IAM configuration and job creation
RR: https://broadworkbench.atlassian.net/browse/PPWM-85
Roughtly 1/3 of storage job creation fails with permission propagation
failures. These effectively stall the pipeline and reduce throughput.
Mitigate the issue by separating updating bucket iam policies and
issuing storage transfer jobs.

[PPWM-86] Retry migrations when GCS is unavailable
RR https://broadworkbench.atlassian.net/browse/PPWM-86
We’ve had a failure in prod where a migration failed while trying to create a new bucket:

  "io.grpc.StatusRuntimeException: UNAVAILABLE: Failed to obtain the location of the GCS
   bucket {bucket} Additional details: GCS is temporarily unavailable."

In such cases, we should retry these later.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

[PPWM-83]: https://broadworkbench.atlassian.net/browse/PPWM-83?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PPWM-84]: https://broadworkbench.atlassian.net/browse/PPWM-84?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PPWM-85]: https://broadworkbench.atlassian.net/browse/PPWM-85?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PPWM-86]: https://broadworkbench.atlassian.net/browse/PPWM-86?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 1]
```diff
@@ -404,26 +403,21 @@ object WorkspaceMigrationActor {
...
-                  billingProject,
-                  // Use a combination of the workspaceId and the current workspace google project id
-                  // as the the resource buffer service (RBS) idempotence token. Why? So that we can
-                  // test this actor with v2 workspaces whose Google Projects have already been claimed
-                  // from RBS via the WorkspaceService.
-                  // The actual value doesn't matter, it just has to be different to whatever the
-                  // WorkspaceService uses otherwise we'll keep getting back the same google project id.
-                  // Adding on the current project id means that this call will be idempotent for all
-                  // attempts at migrating a workspace (until one succeeds, then this will change).
-                  rbsHandoutRequestId = workspace.workspaceId ++ workspace.googleProjectId.value
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:417]: We can't run v2 workspaces down this pipeline anymore so the following is obsolete

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -2178,105 +2179,82 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+  def setupGoogleProjectIam(googleProjectId: GoogleProjectId,
                             policyEmailsByName: Map[SamResourcePolicyName, WorkbenchEmail],
                             billingProjectOwnerPolicyEmail: WorkbenchEmail,
                             span: Span = null): Future[Unit] =
     traceWithParent("updateGoogleProjectIam", span) { _ =>
       logger.info(s"Updating google project IAM ${googleProjectId}.")
-      updateGoogleProjectIam(googleProjectId, policyEmailsByName, terraBillingProjectOwnerRole, terraWorkspaceCanComputeRole, terraWorkspaceNextflowRole, billingProjectOwnerPolicyEmail)
-    }
-
-  private def updateGoogleProjectIam(googleProject: GoogleProjectId,
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:2228]: inline redundant function call

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 3]
```diff
@@ -23,6 +23,7 @@ import org.broadinstitute.dsde.rawls.model._
 import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Implicits._
 import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Outcome.{Failure, Success, toTuple}
 import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils._
+import org.broadinstitute.dsde.rawls.monitor.migration.WorkspaceMigrationActor.MigrateAction._
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:26]: intellij seems to have done this automagically which is quite nice of it. Feel free to ignore changes with the deletion of the `MigrateAction.` lexeme

**Review by aherbst-broad — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 4]
```diff
@@ -404,26 +403,21 @@ object WorkspaceMigrationActor {
...
-                  // as the the resource buffer service (RBS) idempotence token. Why? So that we can
-                  // test this actor with v2 workspaces whose Google Projects have already been claimed
-                  // from RBS via the WorkspaceService.
-                  // The actual value doesn't matter, it just has to be different to whatever the
-                  // WorkspaceService uses otherwise we'll keep getting back the same google project id.
-                  // Adding on the current project id means that this call will be idempotent for all
-                  // attempts at migrating a workspace (until one succeeds, then this will change).
-                  rbsHandoutRequestId = workspace.workspaceId ++ workspace.googleProjectId.value
-                )
+                for {
```
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:409]: Took me a bit to realize this was in an `ifM`, but I understand now 👍

**Comment by rtitle** [comment]
I'm reviewing commit by commit. First 2 commits are 👍 , will pick up the last 2 later this afternoon.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala (hunk 4 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:409]: Yeah - perhaps this is an abuse of `ifM` (though the diff doesn't really help). I welcome any suggestions you may have. I'll admit I've taken a "that'll do" approach and not polish too much as this code shouldn't be too long lived!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala [hunk 5]
```diff
@@ -76,9 +76,12 @@ object FailureModes {
     "%FAILED_PRECONDITION: Service account project-%@storage-transfer-service.iam.gserviceaccount.com " +
       "does not have required permissions%"
 
-  val rateLimitedFailure: String =
+  val stsRateLimitedFailure: String =
     "%RESOURCE_EXHAUSTED: Quota exceeded for quota metric 'Create requests' " +
       "and limit 'Create requests per day' of service 'storagetransfer.googleapis.com'%"
+
+  val gcsUnavailableFailure: String =
+    "%UNAVAILABLE:%Additional details: GCS is temporarily unavailable."
```
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala:84]: Don't think I've ever seen this :)

**Review by rtitle — APPROVED** [review]
Went through the rest, LGTM

**Pull request merged**